src/Microcosm.js

```diff
@@ -20,104 +19,50 @@ let Microcosm = function() {
-   * "released". This lets transactions execute in a predicable order while
-   * not soaking up memory keeping them forever.
-   */
-  this.base = {}
-
-  /**
-   * Holds publically available state. The result of folding all incomplete
-   * transactions over base state. This property can safely be referenced when
-   * retrieving application state.
-   */
-  this.state = this.base
-
+  this.state   = {}
   this.stores  = []
   this.plugins = []
-  this.history = new Tree()
 
+  // Standard store reduction behaviors
   this.addStore(MetaStore)
+
+  // A place to keep all developer warnings
+  this.addPlugin(Debug)
+
+  // Manage key lifecycle hooks for state
+  this.addPlugin(State)
+
+  // Keep track of history as a tree
+  this.addPlugin(History, options)
 }
 
 Microcosm.prototype = {
   constructor: Microcosm,
 
   getInitialState() {
-    return dispatch(this.stores, {}, Transaction(lifecycle.willStart, this.state))
+    return this.lifecycle(lifecycle.willStart)
   },
 
-  /**
-   * Dispatch all open transactions upon base state to determine
-   * a new state. This is the state exposed to the outside world.
-   */
   rollforward() {
-    this.state = this.history.branch().reduce((state, transaction) => {
-      return dispatch(this.stores, state, transaction)
-    }, this.base)
+    this.state = this.lifecycle(lifecycle.willRollforward)
 
     this.emit(this.state)
 
     return this
   },
 
-  shouldHistoryKeep(transaction) {
-    return false
-  },
-
-  clean(transaction) {
-    if (transaction.complete && !this.shouldHistoryKeep(transaction)) {
-      this.base = dispatch(this.stores, this.base, transaction)
-      return true
-    }
-    return false
-  },
-
-  transactionWillOpen(transaction) {
-    this.history.append(transaction)
-  },
-
-  transactionWillUpdate(transaction, error, payload, complete) {
-    transaction.active   = !error
-    transaction.complete = complete
-    transaction.error    = error
-    transaction.payload  = payload
-  },
-
-  transactionWillClose() {
-    this.history.prune(this.clean, this)
+  lifecycle(type, initial={}) {
+    return this.plugins.reduce((state, plugin) => {
+      return plugin[type] ? plugin[type](this, state) : state
+    }, initial)
   },
 
-  /**
-   * Resolves an action. As that action signals changes, it will update
-   * a unique transaction. If an error occurs, it will mark it for clean up
-   * and the change will disappear from history.
-   */
   push(action, params, callback) {
-    if (process.env.NODE_ENV !== 'production' && action == null) {
-      throw new TypeError(`Unable to perform: app.push(${ action })\n\n`
-                          + 'This typically happens when an action is accessed from the wrong key of an object, such as:\n\n'
-                          + '• Actions.mispelledAction\n'
-                          + '• import { mispelledAction } from "actions"')
-    }
+    let transaction = new Transaction(tag(action))
```

Comment:
A transaction in microcosm is a distinct point in time. Every time a transaction is updated, it recalculates state. Application state is the result of dispatching the actions associated with each transaction to a Microcosm's stores. 

This lets you do really smart time travel and have complete predictability on execution order of actions:

![](https://github.com/vigetlabs/microcosm-debugger/raw/master/docs/debugger.gif)

Cool, yeah?